recipes/wpewebkit/0010-Workaround-for-Android-emulator-samplerExternalOES-p.patch

```diff
@@ -0,0 +1,42 @@
+From 9ad6abbe7ce76f3d4bd62d86c3f94ba0fc314f3d Mon Sep 17 00:00:00 2001
+From: Jani Hautakangas <jani@igalia.com>
+Date: Sun, 22 May 2022 23:03:03 +0300
+Subject: [PATCH] Workaround for Android emulator samplerExternalOES precision
+ issue
+
+For some unknown reason glShaderSource fails with 0x505 GL_OUT_OF_MEMORY
+on Android Emulator if precision is not set right after SamplerExternalOESType
+definition
+---
+ .../texmap/TextureMapperShaderProgram.cpp         | 15 +++++++++++++++
+ 1 file changed, 15 insertions(+)
+
+diff --git a/Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp b/Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp
+index 299f0a9e..1f44e1c7 100644
+--- a/Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp
++++ b/Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp
+@@ -593,6 +593,21 @@ Ref<TextureMapperShaderProgram> TextureMapperShaderProgram::create(TextureMapper
+     // Append the common header.
+     fragmentShaderBuilder.append(fragmentTemplateHeaderCommon);
+ 
++    // Workaround for Android emulator samplerExternalOES precision issue.
++    // For some unknown reason glShaderSource fails with 0x505 GL_OUT_OF_MEMORY
++    // on Android Emulator if precision is not set right after SamplerExternalOESType
++    // definition
++    String renderer(reinterpret_cast<const char*>(glGetString(GL_RENDERER)));
++    if (renderer.contains("Android Emulator OpenGL ES Translator")) {
++        String workaround = fragmentShaderBuilder.toString();
++        fragmentShaderBuilder.clear();
++        fragmentShaderBuilder.append(
++                        workaround.replace(
++                            "define SamplerExternalOESType samplerExternalOES",
++                            "define SamplerExternalOESType samplerExternalOES\n"
++                            "precision mediump samplerExternalOES;"));
```

Comment:
There should be a predefined `GL_FRAGMENT_PRECISION_HIGH` macro in GLSL/ESSL that can be used to determine whether `highp` is supported. Ideally we would prefer changing `TextureMapperShaderProgram.cpp` to generate the source for the shader so it becomes:

```glsl
#ifndef GL_FRAGMENT_PRECISION_HIGH
precision mediump samplerExternalOES;
#endif
```

I think this could be done by changing the definition of `OES_EGL_IMAGE_EXTERNAL_DIRECTIVE` right above the fragment shader template header right above. Also, that way the patch could go upstream into WebKit 😸

Reply:
@zhani Or maybe you tried something like that already and I am missing something else that you may know?

Reply:
If I'm not wrong, in OpenGL ES 2.0 the vertex shader sets floats precision to highp by default whereas the fragment shader is NOT REQUIRED to support highp.

So, if you don't explicitly specify floats precision in your shaders you may encounter some issues because they may be specified highp by default by the vertex shader whereas the fragment shader doesn't support it.

This may explain why you encounter this issue with the Android emulator. It is always a good idea to specify explicitly the floats precision at least in the fragment shader using the GL_FRAGMENT_PRECISION_HIGH flag to ensure what is the maximum precision available.

Reply:
works with all precisions lowp, mediump, highp on emulator as long as precision is set right after
"#define SamplerExternalOESType samplerExternalOES", otherwise it fails with out of memory

Reply:
Maybe the emulator doesn't specify any precision by default.

In all cases, maybe it could be better to make the change in the fragment shader itself (the OES_EGL_IMAGE_EXTERNAL_DIRECTIVE define), and not  do a replace() because if names or text content change, it won't work anymore.

Indeed, setting the precision explicitly will work the same on all platforms, it will be harmless outside of Android and even it is part of shaders good practices (https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/OpenGLES_ProgrammingGuide/BestPracticesforShaders/BestPracticesforShaders.html).

Reply:
There's a similar workaround in skia 
https://skia-review.googlesource.com/c/skia/+/235279/

Reply:
Weird thing in our case is that precision needs to be set right after `#define SamplerExternalOESType samplerExternalOES` line. Placing it elsewhere won't work.